httpcore/_utils.py

```diff
@@ -63,3 +64,27 @@ def origin_to_url_string(origin: Origin) -> str:
     scheme, host, explicit_port = origin
     port = f":{explicit_port}" if explicit_port != DEFAULT_PORTS[scheme] else ""
     return f"{scheme.decode('ascii')}://{host.decode('ascii')}{port}"
+
+
+def is_socket_readable(sock_fd: int) -> bool:
+    """
+    Return whether a socket, as identifed by its file descriptor, is readable.
+
+    "A socket is readable" means that it would return immediately with b"" if we
+    called .recv() on it.
+
+    This is also equivalent to "the connection has been closed on the other end".
+
+    See: https://github.com/encode/httpx/pull/143#issuecomment-515181778
+    """
+    # NOTE: We prefer the `selectors` module to `select`, because of known limitations
+    # of `select` on Linux when dealing with many open file descriptors.
+    # See: https://github.com/encode/httpcore/issues/182
+    # On Windows `select` is just fine, but that's also what `DefaultSelector` uses
+    # there, so `selectors` is really the generally-appropriate solution.
+    # See: https://github.com/encode/httpcore/pull/193#issuecomment-703129316
+    sel = selectors.DefaultSelector()
```

Comment:
This will use `epoll` on Linux. From https://github.com/encode/httpcore/pull/193#issuecomment-703129316:
> [...]we're only trying to query a single socket at a time, so the fancy modern options are actually less efficient than good ol' `poll`/`select`.

But I do appreciate that just using `DefaultSelector` makes the code simpler because we don't need to add if-statements for different platforms. So don't really mind 🤷‍♂️ 

OTOH, linking to a comment that suggests _not_ using `selectors` and then going ahead and using it anyway may be a bit confusing.

Reply:
Yup, fair enough, I might just go ahead and steal the trio code that does "select on Windows, poll elsewhere". :)

Reply:
The [comment](https://github.com/encode/httpcore/pull/193#issuecomment-703129316) mentioned by @JayH5 made sense.

I ran local performance tests `selectors` (I'm on Mac) vs `poll`, checking "is_socket_readable" both approaches, so `poll` was faster. After that I aimed google and found some articles like [that](https://daniel.haxx.se/docs/poll-vs-select.html), which told:
> Compared to poll and select, any event-based system will give a performance boost already with a few hundred file descriptors and then the benefit just grows the more connections you add.

So that it's reasonable to copy `trio` approach to our library

Reply:
Yup, swiitched to the `trio` approach!